TiFlash crash under debug mode

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->

```
cmake .. -GNinja -DCMAKE_BUILD_TYPE=Debug -DENABLE_TESTS=OFF
./dbms/src/Server/tiflash server --config-file ../tests/docker/config/tics_dt.toml
```

### 2. What did you expect to see? (Required)

No error happens. Specifically, no error happens when running `tests/delta-merge-tests`

### 3. What did you see instead (Required)

```
[2023/09/11 18:00:27.097 +08:00] [INFO] [Server.cpp:1801] ["Start to wait for terminal signal"] [thread_id=1]
[2023/09/11 18:00:28.570 +08:00] [ERROR] [BaseDaemon.cpp:361] ["(from thread 78) Terminate called after throwing an instance of pingcap::Exception\nCode: 21, e.displayText() = Exception: not implemented, e.what() = Exception\nStack trace:\n\n\n  0x5608e06e628e\tStackTrace::StackTrace() [tiflash+69567118]\n                \tdbms/src/Common/StackTrace.cpp:23\n  0x5608e87e648c\tterminate_handler() [tiflash+204833932]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:628\n  0x7fc807dc2303\tstd::__terminate(void (*)()) [libc++abi.so.1+230147]\n  0x7fc807dc22b8\tstd::terminate() [libc++abi.so.1+230072]\n  0x5608eab6a663\tDB::LocalAdmissionController::startBackgroudJob() [tiflash+242075235]\n                \tdbms/src/Flash/ResourceControl/LocalAdmissionController.cpp:62\n  0x5608e07eb028\tDB::LocalAdmissionController::LocalAdmissionController(pingcap::kv::Cluster*, std::__1::shared_ptr<DB::Etcd::Client>)::'lambda'()::operator()() const [tiflash+70635560]\n                \tdbms/src/Flash/ResourceControl/LocalAdmissionController.h:376\n  0x5608e07eafe5\tdecltype(std::declval<DB::LocalAdmissionControlle"] [source=BaseDaemon] [thread_id=222]
[2023/09/11 18:00:28.570 +08:00] [ERROR] [BaseDaemon.cpp:370] [########################################] [source=BaseDaemon] [thread_id=222]
[2023/09/11 18:00:28.570 +08:00] [ERROR] [BaseDaemon.cpp:371] ["(from thread 78) Received signal Aborted(6)."] [source=BaseDaemon] [thread_id=222]
```

### 4. What is your TiFlash version? (Required)
master

how to reproduce: 
1. tiup cluster start a new cluster with tiflash
2. check tiflash log to wait watchGAC thread start (grep "watchGAC req")
3. kill -9 pd-server_pid
4. we can see tiflash is down, log is:
```
[2023/09/24 11:14:03.464 +08:00] [ERROR] [BaseDaemon.cpp:563] ["\n  0x55e8fdf17e61\tfaultSignalHandler(int, siginfo_t*, void*) [tiflash+216378977]\n                \tlibs/libdaemon/src/BaseDaemon.cpp:214\n  0x7f775bc8bab0\t<unknown symbol> [libc.so.6+236208]\n  0x7f775bcdb26c\t<unknown symbol> [libc.so.6+561772]\n  0x7f775bc8ba08\traise [libc.so.6+236040]\n  0x7f775bc74538\tabort [libc.so.6+140600]\n  0x55e9016511f1\tgrpc::ClientContext::set_call(grpc_call*, std::__1::shared_ptr<grpc::Channel> const&) [tiflash+274285041]\n                \tcontrib/grpc/src/cpp/client/client_context.cc:117\n  0x55e9016551f3\tgrpc::Channel::CreateCallInternal(grpc::internal::RpcMethod const&, grpc::ClientContext*, grpc::CompletionQueue*, unsigned long) [tiflash+274301427]\n                \tcontrib/grpc/src/cpp/client/channel_cc.cc:152\n  0x55e901655309\tgrpc::Channel::CreateCall(grpc::internal::RpcMethod const&, grpc::ClientContext*, grpc::CompletionQueue*) [tiflash+274301705]\n                \tcontrib/grpc/src/cpp/client/channel_cc.cc:160\n  0x55e900c4564e\tgrpc::ClientReaderWriter<etcdserverpb::WatchRequest, etcdserverpb::WatchResponse>::ClientReaderWriter(grpc::ChannelInterface*, grpc::internal::RpcMethod const&, grpc::ClientContext*) [tiflash+263751246]\n                \tcontrib/grpc/include/grpcpp/impl/codegen/sync_stream.h:561\n  0x55e900c27b79\tgrpc::internal::ClientReaderWriterFactory<etcdserverpb::WatchRequest, etcdserverpb::WatchResponse>::Create(grpc::ChannelInterface*, grpc::internal::RpcMethod const&, grpc::ClientContext*) [tiflash+263629689]\n                \tcontrib/grpc/include/grpcpp/impl/codegen/sync_stream.h:439\n  0x55e900be87f5\tetcdserverpb::Watch::Stub::WatchRaw(grpc::ClientContext*) [tiflash+263370741]\n                \tbuild/debug/contrib/etcd/etcd/rpc.grpc.pb.cc:268\n  0x55e90058d39c\tetcdserverpb::Watch::Stub::Watch(grpc::ClientContext*) [tiflash+256705436]\n                \tbuild/debug/contrib/etcd/etcd/rpc.grpc.pb.h:951\n  0x55e90058a682\tDB::Etcd::Client::watch(grpc::ClientContext*) [tiflash+256693890]\n                \tdbms/src/TiDB/Etcd/Client.cpp:210\n  0x55e900575326\tDB::LocalAdmissionController::doWatch() [tiflash+256607014]\n                \tdbms/src/Flash/ResourceControl/LocalAdmissionController.cpp:374\n  0x55e900574e39\tDB::LocalAdmissionController::watchGAC() [tiflash+256605753]\n                \tdbms/src/Flash/ResourceControl/LocalAdmissionController.cpp:354\n  0x55e8f5613a48\tDB::LocalAdmissionController::LocalAdmissionController(pingcap::kv::Cluster*, std::__1::shared_ptr<DB::Etcd::Client>)::'lambda0'()::operator()() const [tiflash+72706632]\n                \tdbms/src/Flash/ResourceControl/LocalAdmissionController.h:390\n  0x55e8f5613a05\tdecltype(std::declval<DB::LocalAdmissionController::LocalAdmissionController(pingcap::kv::Cluster*, std::__1::shared_ptr<DB::Etcd::Client>)::'lambda0'()>()()) std::__1::__invoke[abi:v15007]<DB::LocalAdmissionController::LocalAdmissionController(pingcap::kv::Cluster*, std::__1::shared_ptr<DB::Etcd::Client>)::'lambda0'()>(DB::LocalAdmissionController::LocalAdmissionController(pingcap::kv::Cluster*, std::__1::shared_ptr<DB::Etcd::Client>)::'lambda0'()&&) [tiflash+72706565]\n                \t/usr/bin/../include/c++/v1/__functional/invoke.h:394\n  0x55e8f56139dd\tvoid std::__1::__thread_execute[abi:v15007]<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, DB::LocalAdmissionController::LocalAdmissionController(pingcap::kv::Cluster*, std::__1::shared_ptr<DB::Etcd::Client>)::'lambda0'()>(std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, DB::LocalAdmissionController::LocalAdmissionController(pingcap::kv::Cluster*, std::__1::shared_ptr<DB::Etcd::Client>)::'lambda0'()>&, std::__1::__tuple_indices<>) [tiflash+72706525]\n                \t/usr/bin/../include/c++/v1/thread:284\n  0x55e8f56137ff\tvoid* std::__1::__thread_proxy[abi:v15007]<std::__1::tuple<std::__1::unique_ptr<std::__1::__thread_struct, std::__1::default_delete<std::__1::__thread_struct>>, DB::LocalAdmissionController::LocalAdmissionController(pingcap::kv::Cluster*, std::__1::shared_ptr<DB::Etcd::Client>)::'lambda0'()>>(void*) [tiflash+72706047]\n                \t/usr/bin/../include/c++/v1/thread:295\n  0x7f775bcd944b\t<unknown symbol> [libc.so.6+554059]"] [source=BaseDaemon] [thread_id=233]
```

--- a/dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.h
+++ b/dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.h
@@ -34,9 +34,7 @@ public:
         RUNTIME_CHECK_MSG(
             LocalAdmissionController::global_instance != nullptr,
             "LocalAdmissionController::global_instance has not been initialized yet.");
-        LocalAdmissionController::global_instance->registerRefillTokenCallback([&]() {
-            cv.notify_all();
-        });
+        LocalAdmissionController::global_instance->registerRefillTokenCallback([&]() { cv.notify_all(); });
     }
 
     ~ResourceControlQueue() override { LocalAdmissionController::global_instance->unregisterRefillTokenCallback(); }

--- a/dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.h
+++ b/dbms/src/Flash/Pipeline/Schedule/TaskQueues/ResourceControlQueue.h
@@ -35,7 +35,6 @@ public:
             LocalAdmissionController::global_instance != nullptr,
             "LocalAdmissionController::global_instance has not been initialized yet.");
         LocalAdmissionController::global_instance->registerRefillTokenCallback([&]() {
-            std::lock_guard lock(mu);
             cv.notify_all();
         });
     }

--- a/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
+++ b/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
@@ -324,10 +324,9 @@ std::vector<std::string> LocalAdmissionController::handleTokenBucketsResp(
         {
             LOG_ERROR(
                 log,
-                fmt::format(
-                    "expect resp.granted_r_u_tokens().size() is 1 or 0, but got {} for rg {}",
-                    one_resp.granted_r_u_tokens().size(),
-                    one_resp.resource_group_name()));
+                "expect resp.granted_r_u_tokens().size() is 1 or 0, but got {} for rg {}",
+                one_resp.granted_r_u_tokens().size(),
+                one_resp.resource_group_name());
             continue;
         }
 

--- a/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
+++ b/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
@@ -322,9 +322,12 @@ std::vector<std::string> LocalAdmissionController::handleTokenBucketsResp(
 
         if unlikely (one_resp.granted_r_u_tokens().size() != 1)
         {
-            LOG_ERROR(log, fmt::format(
-                "expect resp.granted_r_u_tokens().size() is 1 or 0, but got {} for rg {}",
-                one_resp.granted_r_u_tokens().size(), one_resp.resource_group_name()));
+            LOG_ERROR(
+                log,
+                fmt::format(
+                    "expect resp.granted_r_u_tokens().size() is 1 or 0, but got {} for rg {}",
+                    one_resp.granted_r_u_tokens().size(),
+                    one_resp.resource_group_name()));
             continue;
         }
 

--- a/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
+++ b/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
@@ -56,13 +56,13 @@ void LocalAdmissionController::startBackgroudJob()
     {
         try
         {
-            // If the unique_client_id cannot be successfully obtained from GAC for a long, then the behavior of resource control is:
+            // If the unique_client_id cannot be successfully obtained from GAC for a long time, then the behavior of resource control is:
             // when the resource group has consumed RU, all queries cannot be scheduled anymore.
             unique_client_id = etcd_client->acquireServerIDFromPD();
         }
         catch (...)
         {
-            // In case we got context timeout error.
+            // Needs to catch in case we got context timeout error.
             LOG_ERROR(
                 log,
                 "get unique_client_id from PD error: {}, resource control may not work properly, try again later",

--- a/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
+++ b/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
@@ -392,7 +392,7 @@ void LocalAdmissionController::watchGAC()
                 return;
 
             // Create new grpc_context for each reader/writer.
-            watch_gac_grpc_context = std::make_shared<grpc::ClientContext>();
+            watch_gac_grpc_context = std::make_unique<grpc::ClientContext>();
         }
     }
 }

--- a/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
+++ b/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
@@ -77,12 +77,10 @@ void LocalAdmissionController::startBackgroudJob()
     auto last_metric_time_point = std::chrono::steady_clock::now();
     while (!stopped.load())
     {
-        std::function<void()> local_refill_token_callback = nullptr;
         bool fetch_token_periodically = false;
 
         {
             std::unique_lock<std::mutex> lock(mu);
-            local_refill_token_callback = refill_token_callback;
 
             auto now = std::chrono::steady_clock::now();
             if (now - last_metric_time_point >= COLLECT_METRIC_INTERVAL)
@@ -107,14 +105,18 @@ void LocalAdmissionController::startBackgroudJob()
             else
                 fetchTokensForLowTokenResourceGroups();
 
-            if (local_refill_token_callback)
-                local_refill_token_callback();
+            {
+                // Need lock here to avoid RCQ has already been destroied.
+                std::lock_guard lock(mu);
+                if (refill_token_callback)
+                    refill_token_callback();
+            }
 
             checkDegradeMode();
         }
         catch (...)
         {
-            handleBackgroundError(getCurrentExceptionMessage(true));
+            LOG_ERROR(log, getCurrentExceptionMessage(true));
         }
     }
 }
@@ -286,7 +288,7 @@ std::vector<std::string> LocalAdmissionController::handleTokenBucketsResp(
 {
     if unlikely (resp.has_error())
     {
-        handleBackgroundError(resp.error().message());
+        LOG_ERROR(log, resp.error().message());
         return {};
     }
 
@@ -297,7 +299,7 @@ std::vector<std::string> LocalAdmissionController::handleTokenBucketsResp(
 
     if (resp.responses().empty())
     {
-        handleBackgroundError("got empty TokenBuckets resp from GAC");
+        LOG_ERROR(log, "got empty TokenBuckets resp from GAC");
         return {};
     }
 
@@ -307,7 +309,7 @@ std::vector<std::string> LocalAdmissionController::handleTokenBucketsResp(
         // For each resource group.
         if unlikely (!one_resp.granted_resource_tokens().empty())
         {
-            handleBackgroundError("GAC return RAW granted tokens, but LAC expect RU tokens");
+            LOG_ERROR(log, "GAC return RAW granted tokens, but LAC expect RU tokens");
             continue;
         }
 
@@ -315,13 +317,17 @@ std::vector<std::string> LocalAdmissionController::handleTokenBucketsResp(
 
         // It's possible for one_resp.granted_r_u_tokens() to be empty
         // when the acquire_token_req is only for report RU consumption.
-        if unlikely (!one_resp.granted_r_u_tokens().empty() && one_resp.granted_r_u_tokens().size() != 1)
+        if (one_resp.granted_r_u_tokens().empty())
+            continue;
+
+        if unlikely (one_resp.granted_r_u_tokens().size() != 1)
         {
-            handleBackgroundError(fmt::format(
+            LOG_ERROR(log, fmt::format(
                 "expect resp.granted_r_u_tokens().size() is 1 or 0, but got {} for rg {}",
                 one_resp.granted_r_u_tokens().size(), one_resp.resource_group_name()));
             continue;
         }
+
         auto resource_group = findResourceGroup(one_resp.resource_group_name());
         if (resource_group == nullptr)
             continue;
@@ -329,7 +335,7 @@ std::vector<std::string> LocalAdmissionController::handleTokenBucketsResp(
         const resource_manager::GrantedRUTokenBucket & granted_token_bucket = one_resp.granted_r_u_tokens()[0];
         if unlikely (granted_token_bucket.type() != resource_manager::RequestUnitType::RU)
         {
-            handleBackgroundError("unexpected request type");
+            LOG_ERROR(log, "unexpected request type");
             continue;
         }
 
@@ -397,7 +403,7 @@ void LocalAdmissionController::doWatch()
     if (!write_ok)
     {
         auto status = stream->Finish();
-        handleBackgroundError(WATCH_GAC_ERR_PREFIX + status.error_message());
+        LOG_ERROR(log, WATCH_GAC_ERR_PREFIX + status.error_message());
         return;
     }
 
@@ -408,13 +414,13 @@ void LocalAdmissionController::doWatch()
         if (!read_ok)
         {
             auto status = stream->Finish();
-            handleBackgroundError(WATCH_GAC_ERR_PREFIX + "read watch stream failed, " + status.error_message());
+            LOG_ERROR(log, WATCH_GAC_ERR_PREFIX + "read watch stream failed, " + status.error_message());
             break;
         }
         LOG_DEBUG(log, "watchGAC got resp: {}", resp.DebugString());
         if (resp.canceled())
         {
-            handleBackgroundError(WATCH_GAC_ERR_PREFIX + "watch is canceled");
+            LOG_ERROR(log, WATCH_GAC_ERR_PREFIX + "watch is canceled");
             break;
         }
         for (const auto & event : resp.events())
@@ -425,11 +431,11 @@ void LocalAdmissionController::doWatch()
             {
             case mvccpb::Event_EventType_DELETE:
                 if (!handleDeleteEvent(kv, err_msg))
-                    handleBackgroundError(WATCH_GAC_ERR_PREFIX + err_msg);
+                    LOG_ERROR(log, WATCH_GAC_ERR_PREFIX + err_msg);
                 break;
             case mvccpb::Event_EventType_PUT:
                 if (!handlePutEvent(kv, err_msg))
-                    handleBackgroundError(WATCH_GAC_ERR_PREFIX + err_msg);
+                    LOG_ERROR(log, WATCH_GAC_ERR_PREFIX + err_msg);
                 break;
             default:
                 RUNTIME_ASSERT(false, log, "unexpect event type {}", magic_enum::enum_name(event.type()));
@@ -515,13 +521,6 @@ bool LocalAdmissionController::parseResourceGroupNameFromWatchKey(
     }
     parsed_rg_name = std::string(etcd_key.begin() + key_prefix.length() + 1, etcd_key.end());
     return true;
-}
-
-void LocalAdmissionController::handleBackgroundError(const std::string & err_msg) const
-{
-    // Basically, errors are all from GAC, cannot handle in tiflash.
-    // So only print log.
-    LOG_ERROR(log, err_msg);
 }
 
 void LocalAdmissionController::checkGACRespValid(const resource_manager::ResourceGroup & new_group_pb)

--- a/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
+++ b/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
@@ -52,15 +52,27 @@ void LocalAdmissionController::warmupResourceGroupInfoCache(const std::string & 
 
 void LocalAdmissionController::startBackgroudJob()
 {
-    try
+    while (!stopped.load())
     {
-        unique_client_id = etcd_client->acquireServerIDFromPD();
+        try
+        {
+            // If the unique_client_id cannot be successfully obtained from GAC for a long, then the behavior of resource control is:
+            // when the resource group has consumed RU, all queries cannot be scheduled anymore.
+            unique_client_id = etcd_client->acquireServerIDFromPD();
+        }
+        catch (...)
+        {
+            // In case we got context timeout error.
+            LOG_ERROR(
+                log,
+                "get unique_client_id from PD error: {}, resource control may not work properly, try again later",
+                getCurrentExceptionMessage(false));
+            std::this_thread::sleep_for(std::chrono::seconds(5));
+            continue;
+        }
+        break;
     }
-    catch (...)
-    {
-        LOG_FATAL(log, getCurrentExceptionMessage(false));
-        std::terminate();
-    }
+    LOG_INFO(log, "get unique_client_id succeed: {}", unique_client_id);
 
     auto last_metric_time_point = std::chrono::steady_clock::now();
     while (!stopped.load())
@@ -363,15 +375,20 @@ void LocalAdmissionController::watchGAC()
         // 2. grpc stream read error.
         // 3. watch is cancel.
         // Will sleep and try again.
-        std::unique_lock lock(mu);
-        if (cv.wait_for(lock, std::chrono::seconds(10), [this]() { return stopped.load(); }))
-            return;
+        {
+            std::unique_lock lock(mu);
+            if (cv.wait_for(lock, std::chrono::seconds(10), [this]() { return stopped.load(); }))
+                return;
+
+            // Create new grpc_context for each reader/writer.
+            watch_gac_grpc_context = std::make_shared<grpc::ClientContext>();
+        }
     }
 }
 
 void LocalAdmissionController::doWatch()
 {
-    auto stream = etcd_client->watch(&watch_gac_grpc_context);
+    auto stream = etcd_client->watch(watch_gac_grpc_context.get());
     auto watch_req = setupWatchReq();
     LOG_DEBUG(log, "watchGAC req: {}", watch_req.DebugString());
     const bool write_ok = stream->Write(watch_req);

--- a/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
+++ b/dbms/src/Flash/ResourceControl/LocalAdmissionController.cpp
@@ -313,11 +313,13 @@ std::vector<std::string> LocalAdmissionController::handleTokenBucketsResp(
 
         handled_resource_group_names.emplace_back(one_resp.resource_group_name());
 
-        if unlikely (one_resp.granted_r_u_tokens().size() != 1)
+        // It's possible for one_resp.granted_r_u_tokens() to be empty
+        // when the acquire_token_req is only for report RU consumption.
+        if unlikely (!one_resp.granted_r_u_tokens().empty() && one_resp.granted_r_u_tokens().size() != 1)
         {
             handleBackgroundError(fmt::format(
-                "expect resp.granted_r_u_tokens().size() is 1, but got {}",
-                one_resp.granted_r_u_tokens().size()));
+                "expect resp.granted_r_u_tokens().size() is 1 or 0, but got {} for rg {}",
+                one_resp.granted_r_u_tokens().size(), one_resp.resource_group_name()));
             continue;
         }
         auto resource_group = findResourceGroup(one_resp.resource_group_name());

--- a/dbms/src/Flash/ResourceControl/LocalAdmissionController.h
+++ b/dbms/src/Flash/ResourceControl/LocalAdmissionController.h
@@ -385,7 +385,7 @@ public:
     LocalAdmissionController(::pingcap::kv::Cluster * cluster_, Etcd::ClientPtr etcd_client_)
         : cluster(cluster_)
         , etcd_client(etcd_client_)
-        , watch_gac_grpc_context(std::make_shared<grpc::ClientContext>())
+        , watch_gac_grpc_context(std::make_unique<grpc::ClientContext>())
     {
         background_threads.emplace_back([this] { this->startBackgroudJob(); });
         background_threads.emplace_back([this] { this->watchGAC(); });
@@ -582,7 +582,7 @@ private:
     ::pingcap::kv::Cluster * cluster = nullptr;
     uint64_t unique_client_id = 0;
     Etcd::ClientPtr etcd_client = nullptr;
-    std::shared_ptr<grpc::ClientContext> watch_gac_grpc_context = nullptr;
+    std::unique_ptr<grpc::ClientContext> watch_gac_grpc_context = nullptr;
     std::vector<std::thread> background_threads;
 
     std::function<void()> refill_token_callback;

--- a/dbms/src/Flash/ResourceControl/LocalAdmissionController.h
+++ b/dbms/src/Flash/ResourceControl/LocalAdmissionController.h
@@ -446,6 +446,9 @@ public:
 
     void registerRefillTokenCallback(const std::function<void()> & cb)
     {
+        // NOTE: Better not use lock inside refill_token_callback,
+        // because LAC needs to lock when calling refill_token_callback,
+        // which may introduce dead lock.
         std::lock_guard lock(mu);
         RUNTIME_CHECK_MSG(refill_token_callback == nullptr, "callback cannot be registered multiple times");
         refill_token_callback = cb;
@@ -516,8 +519,6 @@ private:
     }
 
     std::vector<std::string> handleTokenBucketsResp(const resource_manager::TokenBucketsResponse & resp);
-
-    void handleBackgroundError(const std::string & err_msg) const;
 
     static void checkGACRespValid(const resource_manager::ResourceGroup & new_group_pb);
 

--- a/dbms/src/Flash/ResourceControl/LocalAdmissionController.h
+++ b/dbms/src/Flash/ResourceControl/LocalAdmissionController.h
@@ -385,6 +385,7 @@ public:
     LocalAdmissionController(::pingcap::kv::Cluster * cluster_, Etcd::ClientPtr etcd_client_)
         : cluster(cluster_)
         , etcd_client(etcd_client_)
+        , watch_gac_grpc_context(std::make_shared<grpc::ClientContext>())
     {
         background_threads.emplace_back([this] { this->startBackgroudJob(); });
         background_threads.emplace_back([this] { this->watchGAC(); });
@@ -462,7 +463,13 @@ private:
         if (stopped)
             return;
         stopped.store(true);
-        watch_gac_grpc_context.TryCancel();
+
+        // TryCancel() is thread safe(https://github.com/grpc/grpc/pull/30416).
+        // But we need to create a new grpc_context for each new grpc reader/writer(https://github.com/grpc/grpc/issues/18348#issuecomment-477402608). So need to lock.
+        {
+            std::lock_guard lock(mu);
+            watch_gac_grpc_context->TryCancel();
+        }
         cv.notify_all();
         for (auto & thread : background_threads)
         {
@@ -574,7 +581,7 @@ private:
     ::pingcap::kv::Cluster * cluster = nullptr;
     uint64_t unique_client_id = 0;
     Etcd::ClientPtr etcd_client = nullptr;
-    grpc::ClientContext watch_gac_grpc_context;
+    std::shared_ptr<grpc::ClientContext> watch_gac_grpc_context = nullptr;
     std::vector<std::thread> background_threads;
 
     std::function<void()> refill_token_callback;

